Polyhedron/demo/Polyhedron/Plugins/PMP/Scene_facegraph_item_k_ring_selection.h

```diff
@@ -0,0 +1,667 @@
+#ifndef SCENE_FACEGRAPH_ITEM_K_RING_SELECTION_H
+#define SCENE_FACEGRAPH_ITEM_K_RING_SELECTION_H
+#include "Scene_facegraph_item_k_ring_selection_config.h"
+#ifdef USE_SURFACE_MESH
+#include "Scene_surface_mesh_item.h"
+#include <CGAL/iterator.h>
+#else
+#include "Scene_polyhedron_item.h"
+#include "Polyhedron_type.h"
+#endif
+#include <set>
+#include <QGLViewer/qglviewer.h>
+#include <QKeyEvent>
+#include <QMouseEvent>
+#include <QMainWindow>
+#include <QObject>
+
+#include <map>
+#include <queue>
+
+#include <CGAL/boost/graph/selection.h>
+#include <CGAL/Polygon_mesh_processing/connected_components.h>
+#include <CGAL/Polygon_mesh_processing/border.h>
+
+#include <CGAL/Polygon_2.h>
+
+#ifdef USE_SURFACE_MESH
+typedef Scene_surface_mesh_item Scene_facegraph_item;
+typedef Scene_surface_mesh_item::Kernel FG_Traits;
+#else
+typedef Scene_polyhedron_item Scene_facegraph_item;
+typedef Kernel FG_Traits;
+#endif
+
+typedef Scene_facegraph_item::FaceGraph FaceGraph;
+typedef boost::graph_traits<FaceGraph>::vertex_descriptor vertex_descriptor;
+typedef boost::graph_traits<FaceGraph>::edge_descriptor edge_descriptor;
+typedef boost::graph_traits<FaceGraph>::face_descriptor face_descriptor;
+typedef boost::graph_traits<FaceGraph>::halfedge_descriptor halfedge_descriptor;
+
+struct FG_is_selected_edge_property_map{
+  typedef boost::property_map<FaceGraph,boost::edge_index_t>::type EImap;
+
+  std::vector<bool>* is_selected_ptr;
+  EImap* edge_index_map;
+  FG_is_selected_edge_property_map()
+    : is_selected_ptr(NULL), edge_index_map(NULL) {}
+  FG_is_selected_edge_property_map(std::vector<bool>& is_selected, EImap* map)
+    : is_selected_ptr( &is_selected), edge_index_map(map)
+  {}
+
+  std::size_t id(edge_descriptor ed) {
+    return get(*edge_index_map, ed);
+  }
+
+  friend bool get(FG_is_selected_edge_property_map map, edge_descriptor ed)
+  {
+    CGAL_assertion(map.is_selected_ptr!=NULL);
+    return (*map.is_selected_ptr)[map.id(ed)];
+  }
+
+  friend void put(FG_is_selected_edge_property_map map, edge_descriptor ed, bool b)
+  {
+    CGAL_assertion(map.is_selected_ptr!=NULL);
+    (*map.is_selected_ptr)[map.id(ed)]=b;
+  }
+};
+
+class SCENE_FACEGRAPH_ITEM_K_RING_SELECTION_EXPORT Scene_facegraph_item_k_ring_selection
```

Comment:
qobject_cast is not messed-up, as every plugin that needs is only linked to one and only version of the file. This way there is no conflict. It means the sm version of a plugin will only know the sm version of the sm_item_k_ring_selection, and the poly version will only know the poly_item_k_ring_selection. I agree this is not very clean and this is not easy to decypher, but it works and rethinking the strategy of double targets will take ages and mess-up half a dozen PRs.